Spec.md

```diff
@@ -0,0 +1,240 @@
+
+- An object is mapped to a DataSet.
+- A variable is mapped as a field to the DataSet. If the variable contains children (other variables), a DataSet with the children is created.
+
+#### Example
+
+![image](https://github.com/umati/UA-CloudPublisher/assets/12342602/91eae1d7-9d6f-4246-abb2-0ec5ea9b59bc)
+
+The image shows a part of an address space. Each object is mapped to a DataSet:
+
+- Production
+- ActiveProgram
+- State
+
+The properties Name, NumberInList, Id, and Number are fields of the DataSets. The value of CurrentState is mapped as a field of State and has a DataSet with its properties:
+
+- Production
+- ActiveProgram
+  - Name
+  - NumberInList
+  - CurrentState
+- State
+- CurrentState
+  - Id
+  - Number
+
+### Explanation
+
+A field can also contain properties but no other variables, even though the address space can contain more hierarchies. Properties of objects also cannot be mapped to the properties of a field. Therefore, a uniform mapping is used for objects and variables so that access is always consistent.
+
+## Mapping of References
+
+The name of a DataSet is created by the BrowsePath of the object. The NamespaceUri is used instead of the NamespaceIndex to make the path server-independent. Other BrowsePaths may also be possible. Alternative BrowsePaths can be mapped to the description (see description field of the MetaData). References can also be mapped into the description.
+
+## Mapping of Events
+
+Events are also mapped to a DataSet. Cause Event have no BrowsePath, the BrowsePath of the SourceName and the EventName is used. The mapping itself is analogous to the mapping of objects.
+
+## Topic Structure
+
+The generic topic structure for OPC UA is:
+
+`<Prefix>/<Encoding>/<MqttMessageType>/<PublisherId>/[<WriterGroup>[/<DataSetWriter>]]`
+
+For the umati Dashboard, this topic structure is restricted as follows:
+
+- `<Prefix> = umati.v3`
+- `<Encoding> = json`
+- `<MqttMessageType> = as defined in OPC UA PubSub`
+- `<PublisherId> = Name of the Client`
+- `<WriterGroup> = Expanded NodeId of the object representing the machine`
+- `<DataSetWriter> = BrowsePath`
+
+For identification of the object, the name of the DataSet with expanded NodeId must be used.
+
+The BrowsePath in the DataSetWriter is built with a "." between each BrowseName. The BrowseNames use only the namespace index to avoid collisions.
+All character except `[A-Za-z0-9]` in the `name` field of the BrowseName  need to encode by [URL-Encoding](https://de.wikipedia.org/wiki/URL-Encoding) using an underscore instead of a '%'.
+If two node has the same BrowsePath a iterator ("_Number") can be send to avoid collisions (e.g, `3:Partent.3:Tool_1`, `3:Partent.3:Tool_2` `3:Partent.3:Tool_3` )
+Cause Event have no BrowsePath, the BrowsePath of the SourceName and the EventName is used.
+
+### Examples
+
+DataSet Topic
+
+```text
+umati.v3/json/data/example_publisher_1/nsu=http:_2F_2Fexample.com_2FShowcaseMachineTool_2F;i=66382/2:Identification

```

Comment:
BrowsePaths within Topic's are quite problemeatic...
At least it will not work generic without excluding some reference stuff which is possible in OPC UA but not really in a Topic-String which has a limited length... thats actually a thought one!

Reply:
Maybe we need to restict the topic structure only to hasChild reference. This should solve some issues.